packages/toolpad-app/src/utils/outboundAgent.ts

```diff
@@ -0,0 +1,71 @@
+import agentBase from 'agent-base';
+import http from 'http';
+import https from 'https';
+import dns from 'dns/promises';
+import fetch, { Request } from 'node-fetch';
+
+const SLIDING_WINDOW = 1000;
+const REQS_PER_WINDOW = 5;
+
+const rateLimiterWindows = new Map<string, number[]>();
+
+// Regular garbage collection for the rateLimiterWindows
+setInterval(() => {
+  const now = Date.now();
+  for (const [key, window] of rateLimiterWindows) {
+    if (window.every((timestamp) => timestamp < now - SLIDING_WINDOW)) {
+      rateLimiterWindows.delete(key);
+    }
+  }
+}, 5000).unref();
+
+// Basic sliding window rate limiter
```

Comment:
The way I've used Redis for rate-limiting was with a key that stores a number that gets increment + a certain TTL for the key to expire. I hadn't seen this sliding window method but it should also work, I wonder if it has any advantages compared to the other method i mentioned?

Reply:
For our use-case we're going to want a rate limiting algorithm that can handle bursts. e.g. It's fine to do 3 requests in rapid succession, but not 300. We may also want some cooldown period, e.g. you hit the rate limit, now you're blocked for 60s.

Reply:
Got it, just read about some different rate limiting algorithms, this seems like a good pick.